adds minimal schema to the vdom creation step

mpacer opened the pull request:
This is a WIP; I have not debugged this yet because my vdom version of notebook and nteract broke. But I'll get to that in a sec. 

This should make #3 even easier whoever it is that wants to do that.

mpacer commented:
@rgbkrk I figured that we should try to follow the lead of nbformat and have a directory of schemas so that we can be able to ensure that the right schema works for the right version.

mpacer commented:
Also, I'm going to include a extras_require change in this pr because I realised that we implicitly depend on pytest for testing, but don't declare that anywhere.

rgbkrk commented on CONTRIBUTING.md:
```diff
@@ -38,20 +38,21 @@ would like to address.
-14. Celebrate your success after your pull request is merged! :tada:
+13. Wait for the pull request to be reviewed by a maintainer.
+14. Make changes to the pull request if the reviewing maintainer recommends them.
+15. Celebrate your success after your pull request is merged! :tada:
```
Hey thanks for updating these while you were at it.

rgbkrk commented on schemas/vdom_schema_v0.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "type": "object",
+  "$schema": "http://json-schema.org/draft-04/schema#"
```
Rockin', that's a start.

rgbkrk commented on setup.py:
```diff
@@ -14,6 +14,14 @@
+
+extras_require['all']= list(
+    set([val for k, v in extras_require.items() for val in v])
+    )
```
Thanks!

mpacer commented:
What is the intended use of the schema?

E.g., are we allowing VDOM to accept strings and lists as valid?

I've changed the `to_json` code to actually work through to validate any content that is passed in at the top level (but recursive calls still do not apply validation).

I've also added a bunch of properties to make VDOM itself a more useful class for handing around potential objects while using properties to guarantee a bit more around the validity of the internal structure.

Also, I've added a helper staticmethod for validating VDOM elements without needing to fully instantiate the class (in case at some point in the future that gets to be too much overhead) 

An aside: this is first time I've actually seen a really good use case for a static method! Because it has a default class level schema, it does add value on top of the `to_json` function without having needed to hardcode the reference inside the `to_json` function (which should make it easier to abstract out).

rgbkrk commented on vdom/core.py:
```diff
@@ -18,9 +25,9 @@ def to_json(el, schema=None):
-        return list(map(toJSON, el))
+        json_el = el
+    elif (type(el) is list):
+        json_el = list(map(to_json, el))
```
I'm assuming you moved this into being validated for purity sake. The reason we were short circuiting it before was that it was already valid (since there was no schema, it was the primary invariant here). I think this is fine to send through validation...

rgbkrk replied:
Now that it dawns on me, pure strings and arrays are _not_ valid vdom by themselves (must be a child element). We explicitly have to have an object at the top level due to Jupyter JSON mimetype limitations.

mpacer replied:
It would not be hard to wrap them in a dict if we had a programmatic way of thinking about that… what kinds of strings and Arrays should be allowed? How should they be transformed to objects?

rgbkrk replied:
The top level string could be in a `span`, `div`, `p`, `code`, etc. if we were opinionated on it (note that has an important distinction for layout). An array could be wrapped in a `div`.

However, React handles both `string`s and `Arrays` as top level elements. I'd rather we wait on the next major bump in both the message protocol and notebook format, allowing arrays or strings at the top level.

rgbkrk commented on vdom/core.py:
```diff
@@ -63,12 +70,43 @@ class VDOM():
+
+    @staticmethod
+    def validate(value, schema=_schema):
+        to_json(value, schema=schema)
```
I was thinking the `VDOM` class was pointless prior to this (as we're able to use the raw elements for display). These helpers make it really nice though. Users could also subclass to make updating VDOM output (with updatable displays) really nice.

mpacer replied:
Exactly what I was thinking; especially if people wanted to subclass it for more use-case specific schemas.

rgbkrk replied:
Yeah, what really excited me here was the `obj` property.

rgbkrk replied:
Sometimes I forget that my inline comments make it look like I'm only talking about the line above when sometimes I'm talking about a whole class. Next time I should put my review at the top of the class. 😄

mpacer replied:
then I would just think that you're talking about my choice of what to inherit from 😄.

Also… you should always inherit from object rather than to use an empty pair of parens. That was the source of the py2/py3 bug (as crazy as that sounds).

rgbkrk replied:
Ah that makes sense, I forgot that rule. My Python is rusty.

mpacer replied:
No problem, I'd never run across an issue around importing from object or not before this bug, so I don't think it's terribly common to find yourself in a position where you're forced to attend to that.

rgbkrk commented on vdom/core.py:
```diff
@@ -125,13 +163,23 @@ def __init__(self, *children, **kwargs):
+        @schema.setter
+        def schema(self, value):
+            Draft4Validator.check_schema(value)
+            self._schema = value
```
Now that this has gotten bigger, should we make a `BaseComponent` that we inherit from for our created Component classes?

mpacer replied:
I was thinking that but I didn't want to expand scope!

rgbkrk replied:
Always room to do it after this PR.

rgbkrk commented:
Taking time to respond to your questions / comments now.

> E.g., are we allowing VDOM to accept strings and lists as valid?

Due to limitations with the Jupyter Spec, the top level element of a JSON mimetype `application/whatever+json` must be an object not an array or string. `to_json` should skip those two when figuring out children elements but validate for objects.

> I've also added a bunch of properties to make VDOM itself a more useful class for handing around potential objects while using properties to guarantee a bit more around the validity of the internal structure.

> Also, I've added a helper staticmethod for validating VDOM elements without needing to fully instantiate the class (in case at some point in the future that gets to be too much overhead)

I'm really liking these changes to `VDOM`. It's a nice little helper now. Tempting to call it something like `Layout`, how people will read and write code that uses it.

rgbkrk commented:
> I've changed the to_json code to actually work through to validate any content that is passed in at the top level (but recursive calls still do not apply validation).

This was so not obvious to me at first then I realized it was because of the lack of `schema` argument on the inner calls. It only gets called once for the top level element.

mpacer commented:
>> I've changed the to_json code to actually work through to validate any content that is passed in at the top level (but recursive calls still do not apply validation).
> This was so not obvious to me at first then I realized it was because of the lack of schema argument on the inner calls. It only gets called once for the top level element.

If it's not obvious, should I include a comment saying something to that effect?

rgbkrk commented:
> If it's not obvious, should I include a comment saying something to that effect?

Yeah that'd be cool.

rgbkrk commented:
Since this passes and I think we're pretty happy with it, let's merge this and then we can tackle some of the bits raised in the PR as potential refactors (like the `BaseComponent` class).
